pontoon/base/static/js/translate.js

```diff
@@ -1954,45 +1974,53 @@ var Pontoon = (function (my) {
       });
 
       $('#helpers .history').on('click', 'menu .unapprove', function (e) {
-         var button = $(this),
-             translationId = parseInt($(this).parents('li').data('id'));
+        var button = $(this),
+            translationId = parseInt($(this).parents('li').data('id'));
 
-         $.post('/unapprove-translation/', {
-            csrfmiddlewaretoken: $('#server').data('csrf'),
-            translation: translationId,
-            paths: self.getPartPaths(self.currentPart)
-         }).then(function(data) {
-           var entity = self.getEditorEntity(),
-               pf = self.getPluralForm(true);
-
-           self.stats = data.stats;
-           self.updateProgress(entity);
-
-           self.updateTranslation(entity, pf, data.translation);
-
-           $('#translation').val(data.translation.string).focus();
-           self.updateCachedTranslation();
-           self.updateCurrentTranslationLength();
-
-           if (entity.body && pf === 0) {
-             self.postMessage("SAVE", {
-               translation: data.translation.string,
-               id: entity.id
-             });
-           }
-
-           button.removeClass('unapprove').addClass('approve');
-           button.prop('title', 'Approve');
-           button.parents('li.translated').removeClass('translated').addClass('suggested');
-           button.parents('li').find('.info a').prop('title', self.getApproveButtonTitle({
-             approved: false,
-             unapproved_user: self.user.display_name
-           }));
-
-           self.endLoader('Translation has been unapproved.');
-         }, function() {
-           self.endLoader("Couldn't unapprove this translation.");
-         });
+        $.post('/unapprove-translation/', {
+          csrfmiddlewaretoken: $('#server').data('csrf'),
+          translation: translationId,
+          paths: self.getPartPaths(self.currentPart)
+        }).then(function(data) {
+          var entity = self.getEditorEntity(),
+              pf = self.getPluralForm(true);
+
+          self.stats = data.stats;
+          self.updateProgress(entity);
+
+          self.updateTranslation(entity, pf, data.translation);
+
+          // FTL Editor
+          if ($('#ftl').is('.active')) {
+            self.fluent.renderEditorWithTranslation(data.translation);
+
+          // Standard Editor
+          } else {
+            var translationString = self.fluent.getSimplePreview(data.translation, data.translation.string, entity);
+            $('#translation').val(translationString).focus();
+            self.updateCachedTranslation();
+            self.updateCurrentTranslationLength();
+          }
+
+          if (entity.body && pf === 0) {
+            self.postMessage("SAVE", {
+              translation: data.translation.string,
+              id: entity.id
+            });
+          }
+
+          button.removeClass('unapprove').addClass('approve');
```

Comment:
I hope that React/Angular will free us from that bloat of changing of state of elements inside the js code.